livekit-agents/livekit/agents/pipeline/io.py

```diff
@@ -0,0 +1,211 @@
+
+        self.__nb_playback_finished_needed = 0
+        self.__playback_finished_count = 0
+
+    def on_playback_finished(
+        self, *, playback_position: float, interrupted: bool
+    ) -> None:
+        """
+        Developers building audio sinks must call this method when a playback/segment is finished.
+        Segments are segmented by calls to flush() or clear_buffer()
+        """
+        self.__nb_playback_finished_needed = max(
+            0, self.__nb_playback_finished_needed - 1
+        )
+        self.__playback_finished_count += 1
+        self.__playback_finished_event.set()
+
+        ev = PlaybackFinishedEvent(
+            playback_position=playback_position, interrupted=interrupted
+        )
+        self.__last_playback_ev = ev
+        self.emit("playback_finished", ev)
+
+    async def wait_for_playout(self) -> PlaybackFinishedEvent:
+        """
+        Wait for the past audio segments to finish playing out.
+
+        Returns:
+            PlaybackFinishedEvent: The event that was emitted when the audio finished playing out
+            (only the last segment information)
+        """
+        needed = self.__nb_playback_finished_needed
+        initial_count = self.__playback_finished_count
+        target_count = initial_count + needed
+
+        while self.__playback_finished_count < target_count:
+            await self.__playback_finished_event.wait()
+            self.__playback_finished_event.clear()
+
+        return self.__last_playback_ev
+
+    @property
+    def sample_rate(self) -> int | None:
+        """The sample rate required by the audio sink, if None, any sample rate is accepted"""
+        return self._sample_rate
+
+    @abstractmethod
+    async def capture_frame(self, frame: rtc.AudioFrame) -> None:
+        """Capture an audio frame for playback, frames can be pushed faster than real-time"""
+        if not self.__capturing:
+            self.__capturing = True
+            self.__nb_playback_finished_needed += 1
+
+    @abstractmethod
+    def flush(self) -> None:
+        """Flush any buffered audio, marking the current playback/segment as complete"""
+        if self.__capturing:
+            self.__capturing = False
+
+    @abstractmethod
+    def clear_buffer(self) -> None:
+        """Clear the buffer, stopping playback immediately"""
+        ...
+
+
+class TextSink(ABC):
+    @abstractmethod
+    async def capture_text(self, text: str) -> None:
+        """Capture a text segment (Used by the output of LLM nodes)"""
+        ...
+
+    @abstractmethod
+    def flush(self) -> None:
+        """Mark the current text segment as complete (e.g LLM generation is complete)"""
+        ...
+
+
+# TODO(theomonnom): Add documentation to VideoSink
+class VideoSink(ABC):
+    @abstractmethod
+    async def capture_frame(self, text: rtc.VideoFrame) -> None: ...
+
+    @abstractmethod
+    def flush(self) -> None: ...
+
+
+class AgentInput:
```

Comment:
💯

Reply:
what about text input? ideally there's an interface to hook into the DataStreams API in the same way

Reply:
TextInput can be used with agent.generate_reply("my text")

Reply:
We will need to check for the DataStream API but there is no reason for the text input to be streamed

Reply:
agreed.. the same API is designed for chat as well. but just looking for a similar abstraction.. so it:
- by default, hooks up to a default topic/etc for the DataStreams API 
- allow users to override it too, so that they could hook it up to whatever source